image/png.ksy

```diff
@@ -292,6 +297,56 @@ types:
       - id: text_datastream
         process: zlib
         size-eos: true
+  animation_control_chunk:
+    doc-ref: https://wiki.mozilla.org/APNG_Specification#.60acTL.60:_The_Animation_Control_Chunk
+    seq:
+      - id: num_frames
+        type: u4
+        doc: Number of frames
+      - id: num_plays
+        type: u4
+        doc: Number of times to loop, 0 indicates infinite looping.
+  frame_control_chunk:
+    doc-ref: https://wiki.mozilla.org/APNG_Specification#.60fcTL.60:_The_Frame_Control_Chunk
```

Comment:
The spec lists these constraints that the values have to match:

> Constraints on frame regions:
> 
> ```
>    `x_offset` >= 0
>    `y_offset` >= 0
>    `width`    > 0
>    `height`   > 0
>    `x_offset` + `width`  <= `IHDR` width
>    `y_offset` + `height` <= `IHDR` height
> ```

The first 4 are obviously pointless, because that's implicitly determined by the unsigned-ness of these integers. But it would make sense to check the bottom 2 constraints with the `valid/max` key (kaitai-io/kaitai_struct#435). It is available since the 0.9 release, so be sure to first increment the declared minimum version in [`/meta/ks-version`](https://github.com/kaitai-io/kaitai_struct_formats/blob/4a97145510b2bf3a7dcecf64da46ee363b657e89/image/png.ksy#L18):

```diff
-  ks-version: 0.8
+  ks-version: 0.9
```

I suggest writing the checks like this:

```ksy
      - id: x_offset
        type: u4
        valid:
          max: _root.ihdr.width
        doc: X position at which to render the following frame
```

```ksy
      - id: width
        type: u4
        valid:
          max: _root.ihdr.width - x_offset
        doc: Width of the following frame
```

Of course, it would be actually enough to do the check only once in `width` as the spec suggests (`` `x_offset` + `width`  <= `IHDR` width `` obviously cannot be true if `` `x_offset` > `IHDR` width `` unless `width` is negative, but that can't happen because we're working with unsigned integers). But I think that it's reasonable to report errors on the field that is actually wrong - if there is only a check on `width`, an eventual `ValidationGreaterThanError` will pinpoint the YAML path of `width`, although already `x_offset` might have been invalid.

Reply:
I added a few checks. Unfortunately the web IDE barfs as soon as I add:

```
        valid:
          max: _root.ihdr.width - x_offset
```

Reply:
> Unfortunately the web IDE barfs as soon as I add

Oh, sorry, my bad, I haven't realized that `width` comes before `x_offset`, not after (I was confused by the order of the constraints as presented in the spec). In this case, just do it the other way around:

```ksy
      - id: width
        type: u4
        valid:
          max: _root.ihdr.width
        doc: Width of the following frame
```

```ksy
      - id: x_offset
        type: u4
        valid:
          max: _root.ihdr.width - width
        doc: X position at which to render the following frame
```

It makes actually no real difference - either you consider `x_offset` static and assert that `width` must fit into the remaining space:

```
|            _root.ihdr.width            |
| x_offset | _root.ihdr.width - x_offset |
           | ------> |                   |
              width
```

or you consider `width` static and assert that `x_offset` must adapt:

```
|            _root.ihdr.width            |
| _root.ihdr.width - width |    width    |
| ----------> |            |
   x_offset
```